File: JavaScript/5-BinarySearch/Problems/4-KthSmallestElement-in-a-SortedMatrix/KthSmallestElement-in-a-SortedMatrix.js
/**
 * Given an n x n matrix where each of the rows and columns is sorted in ascending order,
 * return the kth smallest element in the matrix.
 *
 * Note that it is guaranteed that 1 <= k <= n^2.
 *
 *
 * Example 1:
 *
 * Input:
 * matrix = [
 * [ 1,  5,  9],
 * [10, 11, 13],
 * [12, 13, 15]
 * ],
 * k = 8,
 *
 * Output:
 * 13
 *
 * Example 2:
 *
 * Input:
 * matrix = [
 * [ 5,  9],
 * [10, 11],
 * ],
 * k = 3,
 *
 * Output:
 * 10
 *
 *
 * Constraints:
 *
 * n == matrix.length
 * n == matrix[i].length
 * 1 <= n <= 300
 * -10^9 <= matrix[i][j] <= 10^9
 * All the rows and columns of matrix are guaranteed to be sorted in non-decreasing order.
 * 1 <= k <= n^2
 */

/**
 * Approach 1: Brute Force (Merge and Sort)
 * - Merge all elements of the matrix into a single array.
 * - Sort the merged array.
 * - Return the (k-1)th element (since arrays are 0-indexed).
 *
 * Time Complexity: O(N^2 log N^2), where N is the number of rows/columns.
 * Space Complexity: O(N^2), for the merged array.
 */
function kthSmallestBruteForce(matrix, k) {
  const n = matrix.length;
  const mergedArray = [];

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      mergedArray.push(matrix[i][j]);
    }
  }

  mergedArray.sort((a, b) => a - b);
  return mergedArray[k - 1];
}

/**
 * Approach 2: Priority Queue (Min-Heap)
 * - Maintain a min-heap of size k.
 * - Iterate through the matrix, adding elements to the heap.
 * - If the heap size exceeds k, remove the largest element from the heap.
 * - The top element of the heap will be the kth smallest.
 *
 * Time Complexity: O(N^2 log k), where N is the number of rows/columns.
 * Space Complexity: O(k), for the heap.
 */
function kthSmallestPriorityQueue(matrix, k) {
  const n = matrix.length;
  const minHeap = new MinPriorityQueue();

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      minHeap.enqueue(matrix[i][j], matrix[i][j]); //adds element with priority
      if (minHeap.size() > k) {
        minHeap.dequeue(); //removes the min element
      }
    }
  }

  return minHeap.dequeue().element; //returns the kth smallest
}

// Implementation of MinPriorityQueue (Simplified for this problem)
class MinPriorityQueue {
  constructor() {
    this.heap = [];
  }

  enqueue(element, priority) {
    this.heap.push({ element, priority });
    this.bubbleUp(this.heap.length - 1);
  }

  dequeue() {
    if (this.isEmpty()) return null;
    this.swap(0, this.heap.length - 1);
    const popped = this.heap.pop();
    this.bubbleDown(0);
    return popped;
  }

  size() {
    return this.heap.length;
  }
    isEmpty() {
    return this.heap.length === 0;
  }

  bubbleUp(index) {
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      if (this.heap[index].priority >= this.heap[parentIndex].priority) break;
      this.swap(index, parentIndex);
      index = parentIndex;
    }
  }

  bubbleDown(index) {
    while (true) {
      const leftChild = 2 * index + 1;
      const rightChild = 2 * index + 2;
      let smallest = index;

      if (leftChild < this.heap.length && this.heap[leftChild].priority < this.heap[smallest].priority) {
        smallest = leftChild;
      }

      if (rightChild < this.heap.length && this.heap[rightChild].priority < this.heap[smallest].priority) {
        smallest = rightChild;
      }

      if (smallest === index) break;

      this.swap(index, smallest);
      index = smallest;
    }
  }

  swap(i, j) {
    [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
  }
  peek(){
     if (this.isEmpty()) return null;
        return this.heap[0].element;
  }
}

/**
 * Approach 3: Binary Search
 * - The kth smallest element must be within the range [matrix[0][0], matrix[n-1][n-1]].
 * - Perform binary search within this range.
 * - For each mid value, count the number of elements in the matrix that are less than or equal to mid.
 * - Adjust the search range based on whether the count is less than k or greater than or equal to k.
 *
 * Time Complexity: O(N log(max - min)), where N is the number of rows/columns,
 * and max and min are the largest and smallest elements in the matrix.
 * Space Complexity: O(1).
 */
function kthSmallestBinarySearch(matrix, k) {
  const n = matrix.length;
  let low = matrix[0][0];
  let high = matrix[n - 1][n - 1];

  while (low <= high) {
    const mid = low + Math.floor((high - low) / 2);
    let count = 0;
    let j = n - 1;

    for (let i = 0; i < n; i++) {
      while (j >= 0 && matrix[i][j] > mid) j--;
      count += j + 1;
    }

    if (count < k) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return low;
}

/**
 * Approach 4: Optimized Binary Search
 * - Similar to the previous binary search approach, but optimizes the counting.
 * - Instead of counting elements <= mid for each row, we start from the bottom-left.
 * - If the current element is <= mid, we add all elements in the current column (including itself) to the count and move right.
 * - Otherwise, we move up.
 *
 * Time Complexity: O(N log(max - min)), where N is the number of rows/columns.
 * Space Complexity: O(1).
 */
function kthSmallestOptimizedBinarySearch(matrix, k) {
    const n = matrix.length;
    let low = matrix[0][0];
    let high = matrix[n - 1][n - 1];

    while (low <= high) {
        const mid = low + Math.floor((high - low) / 2);
        let count = 0;
        let row = n - 1;
        let col = 0;

        while (row >= 0 && col < n) {
            if (matrix[row][col] <= mid) {
                count += row + 1; // Add the entire column count
                col++;
            } else {
                row--;
            }
        }

        if (count < k) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return low;
}

/**
 * Approach 5: Using a Max Heap of size K
 * - Build a Max Heap of the first row elements.
 * - For the remaining rows, compare the first element of each row with the top of the heap.
 * - If the element is smaller, replace the top of the heap with this element and heapify.
 * - After processing all rows, pop k-1 elements from the heap.
 * - The last element popped is the kth smallest.
 *
 * Time Complexity: O(k log k + (n-1)log k)
 * Space Complexity: O(k)
 */
function kthSmallestMaxHeap(matrix, k) {
  const n = matrix.length;
  const maxHeap = new MaxPriorityQueue();

  // Initialize the heap with the first row
  for (let j = 0; j < n; j++) {
    maxHeap.enqueue({val: matrix[0][j], row: 0, col: j}, matrix[0][j]);
  }

  let result = 0;
  // Pop k-1 times.
  for (let i = 0; i < k; i++) {
    const {val, row, col} = maxHeap.dequeue().element;
    result = val;

    if (row < n - 1) { //if there are more rows to process
      maxHeap.enqueue({val: matrix[row+1][col], row: row+1, col: col}, matrix[row+1][col]);
    }
  }
  return result;
}

// Implementation of MaxPriorityQueue (Simplified for this problem)
class MaxPriorityQueue {
  constructor() {
    this.heap = [];
  }

  enqueue(element, priority) {
    this.heap.push({ element, priority });
    this.bubbleUp(this.heap.length - 1);
  }

  dequeue() {
    if (this.isEmpty()) return null;
    this.swap(0, this.heap.length - 1);
    const popped = this.heap.pop();
    this.bubbleDown(0);
    return popped;
  }

  size() {
    return this.heap.length;
  }
    isEmpty() {
    return this.heap.length === 0;
  }

  bubbleUp(index) {
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      if (this.heap[index].priority <= this.heap[parentIndex].priority) break; // Changed comparison
      this.swap(index, parentIndex);
      index = parentIndex;
    }
  }

  bubbleDown(index) {
    while (true) {
      const leftChild = 2 * index + 1;
      const rightChild = 2 * index + 2;
      let largest = index; // Changed to largest

      if (leftChild < this.heap.length && this.heap[leftChild].priority > this.heap[largest].priority) { // Changed comparison
        largest = leftChild;
      }

      if (rightChild < this.heap.length && this.heap[rightChild].priority > this.heap[largest].priority) { // Changed comparison
        largest = rightChild;
      }

      if (largest === index) break;

      this.swap(index, largest);
      index = largest;
    }
  }

  swap(i, j) {
    [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
  }
  peek(){
     if (this.isEmpty()) return null;
        return this.heap[0].element;
  }
}

// Example usage:
const matrix1 = [
  [1, 5, 9],
  [10, 11, 13],
  [12, 13, 15],
];
const k1 = 8;

const matrix2 = [
  [5, 9],
  [10, 11],
];
const k2 = 3;

console.log("Example 1:");
console.log("Matrix:", matrix1);
console.log("k:", k1);
console.log("Brute Force:", kthSmallestBruteForce(matrix1, k1)); // Output: 13
console.log("Priority Queue:", kthSmallestPriorityQueue(matrix1, k1)); // Output: 13
console.log("Binary Search:", kthSmallestBinarySearch(matrix1, k1)); // Output: 13
console.log("Optimized Binary Search:", kthSmallestOptimizedBinarySearch(matrix1, k1)); // Output: 13
console.log("Max Heap:", kthSmallestMaxHeap(matrix1, k1)); // Output: 13

console.log("\nExample 2:");
console.log("Matrix:", matrix2);
console.log("k:", k2);
console.log("Brute Force:", kthSmallestBruteForce(matrix2, k2)); // Output: 10
console.log("Priority Queue:", kthSmallestPriorityQueue(matrix2, k2)); // Output: 10
console.log("Binary Search:", kthSmallestBinarySearch(matrix2, k2)); // Output: 10
console.log("Optimized Binary Search:", kthSmallestOptimizedBinarySearch(matrix2, k2)); // Output: 10
console.log("Max Heap:", kthSmallestMaxHeap(matrix2, k2)); // Output: 10

File: C++/5-BinarySearch/Problems/4-KthSmallestElement-in-a-SortedMatrix/KthSmallestElement-in-a-SortedMatrix.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <climits>

using namespace std;

// Function to print the matrix (for debugging purposes)
void printMatrix(const vector<vector<int>>& matrix) {
    for (const auto& row : matrix) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }
}

// Approach 1: Using a Max Heap (Priority Queue)
// Time Complexity: O(k log k) -  We add at most k elements to the heap, and heap operations are O(log k).
// Space Complexity: O(k) - The heap stores at most k elements.
int kthSmallest_MaxHeap(const vector<vector<int>>& matrix, int k) {
    if (matrix.empty() || matrix.size() * matrix[0].size() < k) {
        return -1; // Handle invalid input
    }

    priority_queue<int> maxHeap; // Max heap to store the k smallest elements

    for (const auto& row : matrix) {
        for (int val : row) {
            maxHeap.push(val); // Add the current element to the heap
            if (maxHeap.size() > k) {
                maxHeap.pop(); // If the heap size exceeds k, remove the largest element
            }
        }
    }
    return maxHeap.top(); // The top of the max heap is the kth smallest element
}

// Approach 2: Using Binary Search
// Time Complexity: O(n log(maxVal - minVal)) where n is the number of rows/columns. Binary search on the range of values.
// Space Complexity: O(1) - Constant extra space.
int kthSmallest_BinarySearch(const vector<vector<int>>& matrix, int k) {
    if (matrix.empty() || matrix.size() * matrix[0].size() < k) {
        return -1; // Handle invalid input
    }

    int n = matrix.size();
    int low = matrix[0][0];         // Smallest element in the matrix
    int high = matrix[n - 1][n - 1]; // Largest element in the matrix

    while (low <= high) {
        int mid = low + (high - low) / 2; // Calculate the middle value
        int count = 0;                   // Count of elements less than or equal to mid

        // Count elements <= mid
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (matrix[i][j] <= mid) {
                    count++;
                }
                else
                {
                    break; //optimization, since each row is sorted.
                }
            }
        }

        if (count < k) {
            low = mid + 1; // If count is less than k, search in the higher range
        } else {
            high = mid - 1; // If count is greater than or equal to k, search in the lower range
        }
    }
    return low; // 'low' will eventually be the kth smallest element
}

// Approach 3: Using Merge Sort (Merging all rows)
// Time Complexity: O(N log N), where N is the total number of elements in the matrix (n*n).
// Space Complexity: O(N), for the merged array.
int kthSmallest_MergeSort(const vector<vector<int>>& matrix, int k) {
    if (matrix.empty() || matrix.size() * matrix[0].size() < k) {
        return -1;
    }

    vector<int> mergedArray;
    for (const auto& row : matrix) {
        mergedArray.insert(mergedArray.end(), row.begin(), row.end()); //copy all elements.
    }
    sort(mergedArray.begin(), mergedArray.end()); // Sort the merged array

    return mergedArray[k - 1]; // Return the (k-1)th element (0-based indexing)
}

// Approach 4: Optimized Binary Search (Efficient Counting)
// Time Complexity: O(n log(maxVal - minVal)), where n is the number of rows/columns.  Improved counting within the binary search.
// Space Complexity: O(1)
int kthSmallest_OptimizedBinarySearch(const vector<vector<int>>& matrix, int k) {
    if (matrix.empty() || matrix.size() * matrix[0].size() < k) {
        return -1;
    }

    int n = matrix.size();
    int low = matrix[0][0];
    int high = matrix[n - 1][n - 1];

    while (low <= high) {
        int mid = low + (high - low) / 2;
        int count = 0;
        int j = n - 1; // Start from the last column

        for (int i = 0; i < n; ++i) {
            // For each row, find the rightmost element <= mid
            while (j >= 0 && matrix[i][j] > mid) {
                j--;
            }
            count += (j + 1); // Add the count of elements <= mid in this row
        }

        if (count < k) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return low;
}

// Approach 5: Using a Min Heap (Priority Queue) - Optimized for Sorted Matrix
// Time Complexity: O(k log n) -  We add at most k elements to the heap, and heap operations are O(log n).  n is the dimension of matrix.
// Space Complexity: O(n) - The heap stores at most n elements (one from each row).
int kthSmallest_MinHeap(const vector<vector<int>>& matrix, int k) {
    if (matrix.empty() || matrix.size() * matrix[0].size() < k) {
        return -1;
    }

    int n = matrix.size();
    // Use a min heap to store elements of the matrix.  The pair is (value, {row, col}).
    priority_queue<pair<int, pair<int, int>>, vector<pair<int, pair<int, int>>>, greater<pair<int, pair<int, int>>>> minHeap;

    // Push the first element of each row into the min heap.
    for (int i = 0; i < n; ++i) {
        minHeap.push({matrix[i][0], {i, 0}});
    }

    int count = 0;
    int result = 0;
    while (!minHeap.empty() && count < k) {
        pair<int, pair<int, int>> current = minHeap.top();
        minHeap.pop();
        result = current.first; // Get the smallest element
        count++;

        int row = current.second.first;
        int col = current.second.second;

        if (col + 1 < n) {
            // Push the next element from the same row into the heap
            minHeap.push({matrix[row][col + 1], {row, col + 1}});
        }
    }
    return result;
}

int main() {
    vector<vector<int>> matrix = {
        {1,  5,  9},
        {10, 11, 13},
        {12, 13, 15}
    };
    int k = 8;

    cout << "Matrix:" << endl;
    printMatrix(matrix);
    cout << "k = " << k << endl << endl;

    cout << "1. Using Max Heap: " << kthSmallest_MaxHeap(matrix, k) << endl;
    cout << "2. Using Binary Search: " << kthSmallest_BinarySearch(matrix, k) << endl;
    cout << "3. Using Merge Sort: " << kthSmallest_MergeSort(matrix, k) << endl;
    cout << "4. Using Optimized Binary Search: " << kthSmallest_OptimizedBinarySearch(matrix, k) << endl;
    cout << "5. Using Min Heap: " << kthSmallest_MinHeap(matrix, k) << endl;

    return 0;
}
